File: solutions/elixir/matching-brackets/lib/matching_brackets.ex
defmodule MatchingBrackets do
  @moduledoc false
  @open ~w([ { \()
  @close ~w(] } \))
  @match %{"[" => "]", "{" => "}", "(" => ")"}

  @doc """
  Checks that all the brackets and braces in the string are matched correctly, and nested correctly
  """
  @spec check_brackets(String.t()) :: boolean
  def check_brackets(str), do: closers_match?(str, [])

  @doc false
  @spec closers_match?(String.t(), [String.t()]) :: boolean
  defp closers_match?("", []), do: true
  defp closers_match?("", _), do: false

  defp closers_match?(<<first::binary-size(1), rest::binary>>, closers) do
    cond do
      first in @close and closers == [] -> false
      first in @close and hd(closers) != first -> false
      first in @close -> closers_match?(rest, tl(closers))
      first in @open -> closers_match?(rest, [@match[first] | closers])
      true -> closers_match?(rest, closers)
    end
  end
end

File: solutions/clojure/matching-brackets/src/matching_brackets.clj
(ns matching-brackets)

(def ^:private pairs {\[ \], \{ \}, \( \)})
(def ^:private opener (-> pairs keys set))
(def ^:private closer (-> pairs vals set))

(defn- closers-match? [remaining closers]
  (let [[next & the-rest] remaining]
    (cond
      ;; if no characters are left, check that no more closers are expected
      (empty? remaining) (empty? closers)
      ;; if next character is a closer, check it matches the latest expected closer
      (contains? closer next) (let [[expected & the-others] closers]
                                (if (= next expected) (closers-match? the-rest the-others) false))
      ;; if next character is an opener, add corresponding value to expected closers
      (contains? opener next) (closers-match? the-rest (-> pairs (get next) (cons closers)))
      ;; otherwise, check the rest of the characters for any expected closers
      :else (closers-match? the-rest closers))))

(defn valid?
  "Checks that all the brackets and braces in the input are matched correctly, and nested correctly"
  [input]
  (-> input seq (closers-match? [])))